### DISCUSSION TITLE
FMU export from OpenModelica to Python (FMPy)

### DISCUSSION THREAD

**Discussion by LeoSchmider** [OpenModelica/OpenModelica#12018] [Q&A]
@AnHeuermann  and I, where discussing (from [here](https://stackoverflow.com/questions/78019725/fmu-export-from-openmodelica-to-python-fmpy/)) how to export FMU from OpenModelica (v1.23.dev from (20/02/2023)) in 2.0 and Co-simulation in Windows 11 64 bits using the option native from OpenModelica and added Simulation tag (--fmuRuntimeDepends=modelica). The model I tried to export was the basic pendulum. When I try to open the file .fmu from FMPy GUI (v0.3.20) and Python (3.12.2), I've got the following error : "Failed to load shared library"

C:\...\Temp\tmpbwoddm6i\binaries\win64\FMU_pendulum.dll. Could not find module C:\...\Temp\tmpbwoddm6i\binaries\win64\FMU_pendulum.dll (or one of its dependencies).

**Comment by AnHeuermann** [comment]
Do you need 32bit FMUs? Let's start by investigating the 64 Windows build only (only native platform, no Docker cross-compilation).

You can check what library is missing by using tools like [Dependencies](https://github.com/lucasg/Dependencies). I used it to check what's missing in https://github.com/OpenModelica/OpenModelica/issues/12013#issuecomment-1954658731.

Unzip the FMU, open the `FMU/binaries/win64/<modelname>.dll` and check what dll can't be found.
What files do you have in your win64 directory?

**Comment by LeoSchmider** [comment]
Well I might need them later on, but i'm okay to first go through the bug with the win64 version.
When I drag and drop the .dll file in the GUI of Dependencies, it tells me that libwinpthread-1.dll is the problem. I have a .dlla file too should I run it through Dependencies ?

**Reply by LeoSchmider** [reply]
Totally unrelated but I have just noticed it on the OMEdit frontpage the is a typo with the new version update from yesterday it's the 1.22.2 the typo is on the official webpage too.

**Reply by AnHeuermann** [reply]
The `libwinpthread-1.dll` should be right next to the `<modelname>.dll`. This is what `--fmuRuntimeDepends=modelica` will do. On my machine this is also working with the default value `--fmuRuntimeDepends=default` (when using the changes from https://github.com/OpenModelica/OpenModelica/pull/12014).

What happens if you run the following script with `omc`?

First create file `Pendulum.mo`

```
model Pendulum
  extends Modelica.Mechanics.MultiBody.Examples.Elementary.Pendulum;
  annotation(
    uses(Modelica(version="4.0.0")),
    experiment(StartTime=0.0, StopTime=5));
end Pendulum;
```

and file `test.mos`

```
loadFile("Pendulum.mo"); getErrorString();
setCommandLineOptions("--fmuRuntimeDepends=modelica"); getErrorString();
buildModelFMU(Pendulum); getErrorString();
```

Start a terminal from OMEdit (Tools->OpenModelica Command Prompt)

![grafik](https://github.com/OpenModelica/OpenModelica/assets/38031952/8ccbe135-327b-477e-a719-2945a450fe44)

and run
```cmd
> cd C:/path/to/test.mos
> omc test.mos
```

Will this add dependency `libwinpthread-1.dll` in your `win64` directory?

**Reply by perost** [reply]
> Totally unrelated but I have just noticed it on the OMEdit frontpage the is a typo with the new version update from yesterday it's the 1.22.2 the typo is on the official webpage too.

The webpage is fixed now, thanks for telling us. And OMEdit just pulls the news from the webpage.

**Reply by LeoSchmider** [reply]
You're welcome :)

**Reply by LeoSchmider** [reply]
This is the result I get from the command window but the fmu file is the same, I mean the win64 folder has the same two file (Pendulum.dll and libPendulum.dll.a) and when I drag and drop the Pendulum.dll file in Depencies the same file is missing.

![Screenshot_from_omc](https://github.com/OpenModelica/OpenModelica/assets/150122360/3da769be-7b6c-48e8-b455-104316b6afe5)

**Reply by AnHeuermann** [reply]
Ignore the warnings (typos in Pendulum.mo).

That's strange. My setup is a bit different, since I have all of the developer tools in my `PATH`.
What OpenModelica version are you using (with hash / dev number)? Do you have more than one OpenModelica installed?

Can you upload the FMU here (change .fmu to .zip)?

I can try to create a clean OpenModelica installation inside a Windows Docker environment and see if this is missing the dependency as well.

**Comment by LeoSchmider** [comment]
I'm using the OpenModelica v1.23.-dev-324-ga10c18c83b (64-bits). 
Well, every time I download a new one the installer has a phase where it deletes all the previous versions files.
Sure, I'm not sure how to do the conversion from .fmu to .zip if you have any tips. I am just going to put the .fmu in another .zip file.
[Pendulum.zip](https://github.com/OpenModelica/OpenModelica/files/14362441/Pendulum.zip)

**Reply by AnHeuermann** [reply]
I can see that the setting from `fmuRuntimeDepends` did end up in the `FMU/sources/CMakeLists.txt`. Do you have file `C:/Program Files/OpenModelica1.23.0-dev-64bit/bin/libwinpthread-1.dll` in that exact location?

When I re-compile your FMU with the CMakeLists.txt file pthread is added to the bin directory:

Start cmd from OMEdit again:

```cmd
> cmake --version
> cd C:\path\to\Pendulum\sources\
> cmake -S . -B build_cmake -Wno-dev -G "MSYS Makefiles" 
> cd build_cmake
> make install -j -Oline
```

with output

```
[...]
[100%] Linking C shared library Pendulum.dll
[100%] Built target Pendulum
Install the project...
-- Install configuration: ""
-- Installing: D:/Pendulum/Pendulum/sources/../binaries/win64/libPendulum.dll.a  
-- Installing: D:/Pendulum/Pendulum/sources/../binaries/win64/Pendulum.dll       
-- Installing: D:/Pendulum/Pendulum/sources/../binaries/win64/libwinpthread-1.dll
```

I don't get what's going wrong on your side. Maybe try again with tomorrows nightly build?
If that fails I'll try it again in a clean VM.

**Reply by LeoSchmider** [reply]
I actually do have the file C:/Program Files/OpenModelica1.23.0-dev-64bit/bin/libwinpthread-1.dll in that location.
Ok thank you for your help, I will post you here the result of my attempt with tomorrow nighty build.

**Reply by LeoSchmider** [reply]
Hi, I've tried with today's version of OpenModelica but the bug is still present.

**Reply by AnHeuermann** [reply]
I created a virtual machine with Windows 11 developer environment using the Microsoft Hyper-V manager.

Then I installed OpenModelica, exported `Modelica.Mechanics.MultiBody.Examples.Elementary.Pendulum` with the default settings and the resulting FMU has a copy of pthreads in the binaries directory.

![grafik](https://github.com/OpenModelica/OpenModelica/assets/38031952/c6e47b31-cd21-4927-8601-a461a07b4fdb)

Sorry, I can't reproduce your issue. With `v1.23.0-dev-337-g33535cb75a` this works out of the box, no need to change any settings.
Maybe you still have some settings different to the default. You can try to reset your OMEdit settings in Tools->Options and hit the `Reset` button.

Or there is some other stuff going on you didn't notice.

**Reply by LeoSchmider** [reply]
Just tried to reset but the problem still persists, I really don't know what to tell you. If you have any idea of like a footage of whatever manipulation or anything to help, you can tell me.

**Reply by AnHeuermann** [reply]
I think there isn't much I can do at this point. I can't setup a Windows 10 virtual machine to check if it's working there as well.

If you want to spend some more time on this and you are familiar with compiling C and CMake you can try to re-compile your FMU (FMU/sources/CMakeLists.txt) and try to find an error message or warning in the output. The compilation process is described in https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/fmitlm.html#compilation-process.


**Workaround:**
If you really need that FMU to work you can manually copy `libwinpthread-1.dll` into the win64 directory and re-zip the FMU.

**Reply by LeoSchmider** [reply]
Thank you so much for your time and help. I have tried on a machine with windows 10 and the problem is still present for me. I have discovered something interesting I don't know if it is related to the same file (libwinpthread-1.dll, I don't currently access to the machine, so when I can I'm going to download Dependencies and I'll get back to you) but when I export an FMU file from OpenModelica 1.18.0 in 32bits the FMU works, but with OpenModelica 1.22.1 in 32 bits the FMU doesn't work. The only path I have to check for some errors in that the function fmi2Status gives the result -2^31.  (both files where exported in native platform without using docker)

**Reply by AnHeuermann** [reply]
Wait, are you on a 32bit Windows 10 OS?

**Reply by LeoSchmider** [reply]
I have 3 setups, I have one windows 11 64 bits and one windows 10 64 bits where I have the newest OpenModelica and exported with native setup and tried to open the fmu with python (FMPy)and on another hand I have a vm with windows 10 64 bits but I downloaded both Openmodelica 1.22 32 bits and 1.18.0 32 bits and exported fmu as native to open the file with another program called simit cte ( which requires FMU 2.0, CS, and win32). I have found out the fmu generated with openmodelica 1.18.0 works in simit cte and but the one generated with openmodelica 1.22 doesn't. For now I don't know if it is related to the libwinpthread-1.dll file but I'm going to check tomorrow. Sorry if the message above wasn't clear.

**Reply by AnHeuermann** [reply]
Sorry I'm lost. You are using three setups with two different OpenModelica versions to compile FMUs for three different platforms and combine this with Docker. No chance to reproduce this.

If you are still having issues with the latest OpenModelica nightly build (v.1.23-dev) on a 64bit Windows OS you might want to open an issue and explain to us how to reproduce your issue using MOS scripts like the one I posted somewhere above.
If I can reproduce your issue chances are high I'll fix it eventually.

**Reply by LeoSchmider** [reply]
Okay I'll try to be a bit clearer. I have tried to export my fmus from openmodelica 1.22 (from win 11 and win 10 both 64 bits to check if it was a windows version problem) to open them with simit cte, but simit only opens win32 fmus so I used docker. It didn't seem to work. So on the fmi standard website I have found that there was a python tool to try to open my fmus (FMPy GUI), but it failed too, that's when I tried to ask for help on stackoverflow (you know the rest of the story). For now I don't know how to solve the problem with FMPy GUI that we have identified but I have a workaround for simit cte which is to download openmodelica 1.18.0 32 bits and the fmus works fine (to make sure the problem was not docker related I tried to export the fmus from openmodelica 1.22 32 bits and it still did not work). I'm not sure if the bug is the same for FMPy GUI and Simit (I'll check tomorrow with dependencies and come back to you) but the explanations were merely to make it easier to reproduce the bugs.
For FMPy GUI, the problem seems to come from libwinpthread-1.dll (independently of the os or the openmodelica version I use). For Simit cte the only problem I can get the software to tell me is that the FMI2state = -2^31. For both tools the fmu generation process is the same in openmodelica. (Create the file Pendulum.mo, write the code, right click on the model in the library window, export -> fmu) the setup for fmu export is the mostly the same Fmu 2.0, CS, and the platform is native for FMPy and i686-w64-ming32 for simit on the newest version of openmodelica. For v1.18 I only put static for the platform. I run Openmodelica, Docker, Simit and FMPy as admin.

**Reply by casella** [reply]
@LeoSchmider, I would suggest you to try the latest 1.23.0-dev nightly build for win64 and cross-compile to win32, if you need 32-bit FMUs

**Reply by LeoSchmider** [reply]
Update : I've tried to export the pendulum FMU file from OpenModelica 1.18 32 bits and analysed it with dependencies and no file is missing so, I do believe the non compatibility with simit cte might come from the missing .dll file in posterior versions of openmodelica (same problem than FMPy). I have opened it with the lastest version of FMPy and it works perfecty fine.

**Reply by LeoSchmider** [reply]
@casella Do you mean to generate fmus with the win32 spec from omedit or to use native spec and then reprocess it with cmake ?

**Reply by casella** [reply]
The former

**Reply by LeoSchmider** [reply]
Sorry, but I'm kind of new to FMUs, could you be a bit more specific, please.
